File: src/net/dinkla/raytracer/tracers/Tracer.java
package net.dinkla.raytracer.tracers;

import net.dinkla.raytracer.colors.Color;
import net.dinkla.raytracer.math.Ray;
import net.dinkla.raytracer.math.WrappedFloat;
import net.dinkla.raytracer.worlds.World;
import org.apache.log4j.Logger;

abstract public class Tracer {

    /**
     * The logger.
     */
    static final Logger LOGGER = Logger.getLogger(Tracer.class);

    /**
     * The world.
     */
    public World world;

    /**
     * Constructor.
     *
     * @param world
     */
    public Tracer(World world) {
        this.world = world;
    }

    // TODO: Warum so viele trace-Funktionen?
    // TODO: Sollte nicht die mit den meisten Parametern abstract sein?
    abstract public Color trace(Ray ray, final int depth);

    public Color trace(Ray ray, WrappedFloat tmin, int depth) {
        //LOGGER.debug("trace " + ray + " tmin=" + tmin + " at depth " + depth);
        return trace(ray, depth);
    }

    public Color trace(Ray ray) {
        //LOGGER.debug("trace " + ray);
        return trace(ray, 0);
    }    
}

File: src/main/kotlin/net/dinkla/raytracer/tracers/Tracer.kt
package net.dinkla.raytracer.tracers

import net.dinkla.raytracer.colors.Color
import net.dinkla.raytracer.math.Ray
import net.dinkla.raytracer.math.WrappedDouble
import net.dinkla.raytracer.world.World
import org.slf4j.LoggerFactory

interface Tracer {

    abstract fun trace(ray: Ray, depth: Int): Color

    open fun trace(ray: Ray, tmin: WrappedDouble, depth: Int): Color = trace(ray, depth)

    open fun trace(ray: Ray): Color = trace(ray, 0)

}
